File: clients/kotlin/src/main/kotlin/streamsforge/GrpcTransport.kt
package streamsforge

import com.google.protobuf.Empty
import io.grpc.ManagedChannel
import io.grpc.ManagedChannelBuilder
import io.grpc.Metadata
import io.grpc.netty.shaded.io.grpc.netty.GrpcSslContexts
import io.grpc.netty.shaded.io.grpc.netty.NettyChannelBuilder
import io.grpc.netty.shaded.io.netty.handler.ssl.util.InsecureTrustManagerFactory
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.flow.flowOf
import kotlinx.coroutines.flow.map
import java.io.File
import streamsforge.v1.GetTableRowsRequest
import streamsforge.v1.IngestRequest
import streamsforge.v1.IngestServiceGrpcKt
import streamsforge.v1.StreamServiceGrpcKt
import streamsforge.v1.SubscribeTableRequest
import streamsforge.v1.TableServiceGrpcKt

private val AUTH_KEY: Metadata.Key<String> = Metadata.Key.of("authorization", Metadata.ASCII_STRING_MARSHALLER)

/** One parsed `grpcTarget`: [host]/[port] plus whether it asked for TLS. See [parseGrpcTarget]. */
internal data class GrpcAddress(val host: String, val port: Int, val tls: Boolean)

/**
 * Parses the three shapes [StreamsForge.connect]'s `grpcTarget` accepts:
 * - `host:port` -- gRPC's own scheme-less target syntax, plaintext, unchanged from before TLS
 *   support existed.
 * - `http://host:port` -- plaintext, spelled out explicitly (e.g. by `StreamsForge`'s own
 *   `defaultGrpcTarget` guess when the REST `url` itself was `http://`).
 * - `https://host:port` -- TLS, ALPN h2 (matches a host started with `--Tls:Enabled true`).
 *
 * A bare target's port is required (gRPC's own convention); a schemed target without an explicit
 * port falls back to the scheme's standard port (80/443) -- matching how a browser would resolve
 * the same URL, even though StreamsForge hosts always bind an explicit port in practice.
 */
internal fun parseGrpcTarget(target: String): GrpcAddress = when {
    target.startsWith("https://") -> splitHostPort(target.removePrefix("https://"), 443).let { (h, p) -> GrpcAddress(h, p, tls = true) }
    target.startsWith("http://") -> splitHostPort(target.removePrefix("http://"), 80).let { (h, p) -> GrpcAddress(h, p, tls = false) }
    else -> splitHostPort(target, -1).let { (h, p) ->
        require(p != -1) { "grpcTarget '$target' has no port -- expected host:port, http://host:port or https://host:port" }
        GrpcAddress(h, p, tls = false)
    }
}

private fun splitHostPort(hostPort: String, defaultPort: Int): Pair<String, Int> {
    val idx = hostPort.lastIndexOf(':')
    return if (idx < 0) hostPort to defaultPort else hostPort.substring(0, idx) to hostPort.substring(idx + 1).toInt()
}

/**
 * Plaintext (`host:port` / `http://host:port`) goes through the SAME `ManagedChannelBuilder.
 * forTarget(...).usePlaintext()` call as before TLS support existed -- an h2c channel, prior
 * knowledge, matching how the engine runs from source. `https://host:port` switches to
 * `NettyChannelBuilder.forAddress` with ALPN h2 (`useTransportSecurity()`), trusting [caFile] (the
 * dev cert IS its own CA -- see `tools/tls/dev-cert.sh`) or, for [insecure], every certificate via
 * grpc-netty-shaded's own `InsecureTrustManagerFactory` -- development only, never the default.
 *
 * Note this builds a grpc-netty `SslContext`, NOT the `javax.net.ssl.SSLContext` [buildTlsConfig]
 * produces for REST/SignalR -- the two SSL stacks don't share a type, so gRPC gets its own
 * construction from the same raw `caFile`/`insecure` inputs (see [TlsConfig]'s doc for why).
 */
private fun buildChannel(address: GrpcAddress, caFile: String?, insecure: Boolean): ManagedChannel {
    if (!address.tls) {
        return ManagedChannelBuilder.forTarget("${address.host}:${address.port}").usePlaintext().build()
    }
    val sslContextBuilder = GrpcSslContexts.forClient()
    when {
        insecure -> sslContextBuilder.trustManager(InsecureTrustManagerFactory.INSTANCE)
        caFile != null -> sslContextBuilder.trustManager(File(caFile))
        // Neither insecure nor a CA given for an https:// target: fall through to the JVM's
        // default trust store, same as any other TLS client -- correct for a host with a
        // CA-issued certificate, and the caller's job to have supplied caFile/insecure otherwise.
    }
    return NettyChannelBuilder.forAddress(address.host, address.port)
        .useTransportSecurity()
        .sslContext(sslContextBuilder.build())
        .build()
}

/** Result of one `IngestService.Ingest` bidi push (design doc §3.1: real HTTP/2 backpressure --
 * the server does not ack until the push is admitted -- instead of REST's `retry_after_ms`
 * guess). Mirrors `StreamsForge.Abstractions.IngestOutcome` (`IngestAck` in streamsforge.proto). */
data class IngestAckResult(
    val outcome: String,
    val accepted: Int,
    val dropped: Int,
    val invalid: Int,
    val retryAfterMs: Int,
    val error: String,
    val rowErrors: List<String>,
    val duplicate: Int,
    val replayed: Boolean,
)

/**
 * Tier 1 gRPC transport (design doc §3): `StreamService.SubscribeTable` for deltas,
 * `TableService.Rows/List` for the catalog and snapshot, bidi `IngestService.Ingest` for pushes.
 * One insecure h2c channel, prior knowledge -- matching how the engine actually runs from source
 * (§3.2's `--urls` trap: it must be started with `--Http:Port`/`--Grpc:Port`, never `--urls`, or
 * no gRPC port is bound at all).
 *
 * Every gRPC service carries `[Authorize(Policy = "Viewer")]`, so an `authorization: Bearer <jwt>`
 * metadata entry is all that's needed -- [tokenProvider] is `AuthClient::token`, shared with the
 * REST side so both transports mint/refresh the same JWT (design doc §3.1).
 */
class GrpcTransport(
    target: String,
    caFile: String? = null,
    insecure: Boolean = false,
    private val tokenProvider: suspend () -> String,
) : TableTransport {
    override val name = "grpc"

    private val channel: ManagedChannel = buildChannel(parseGrpcTarget(target), caFile, insecure)
    private val tableStub = TableServiceGrpcKt.TableServiceCoroutineStub(channel)
    private val streamStub = StreamServiceGrpcKt.StreamServiceCoroutineStub(channel)
    private val ingestStub = IngestServiceGrpcKt.IngestServiceCoroutineStub(channel)

    private suspend fun authHeaders(): Metadata = Metadata().apply { put(AUTH_KEY, "Bearer ${tokenProvider()}") }

    override fun close() {
        channel.shutdownNow()
    }

    /** Proves the channel AND the JWT actually work -- used by `StreamsForge.connect(transport =
     * AUTO)` to decide whether gRPC is reachable before committing to it (design doc: `AUTO`
     * "tries gRPC, falls back to SignalR, and logs which one it got"). */
    suspend fun probe() {
        tableStub.list(Empty.getDefaultInstance(), authHeaders())
    }

    private suspend fun resolveTableId(tableName: String): String {
        val resp = tableStub.list(Empty.getDefaultInstance(), authHeaders())
        return resp.tablesList.firstOrNull { it.name == tableName }?.id
            ?: throw StreamsForgeError("no such table '$tableName'")
    }

    override suspend fun snapshot(tableName: String, limit: Int): Pair<List<RowDelta>, Long> {
        val id = resolveTableId(tableName)
        val request = GetTableRowsRequest.newBuilder().setId(id).setLimit(limit).build()
        val response = tableStub.rows(request, authHeaders())
        return response.rowsList.map { RowDelta(it.row.toRow(), it.weight) } to response.seq
    }

    override suspend fun subscribeTable(tableName: String): Flow<DeltaBatch> {
        // Metadata (the bearer token) is fetched HERE, as part of this suspend call, not lazily
        // inside a `flow {}` builder -- see [TableTransport.subscribeTable]'s doc: a cold Flow
        // that defers even cheap prep work to collection time can let the caller's REST snapshot
        // race ahead of the subscription actually being registered. gRPC's own stream setup
        // (`streamStub.subscribeTable`) stays a plain non-suspend call returning a Flow, same as
        // grpc-kotlin generates it -- fetching the token first is what matters here, since the
        // token itself refreshes on every (re)subscribe, mirroring live.py's "every reconnect
        // starts fresh".
        val headers = authHeaders()
        val request = SubscribeTableRequest.newBuilder().setName(tableName).build()
        return streamStub.subscribeTable(request, headers).map { batch ->
            DeltaBatch(batch.deltasList.map { RowDelta(it.row.toRow(), it.weight) }, batch.seq)
        }
    }

    /** One request, one ack, over a fresh bidi stream -- gets real backpressure semantics from
     * the bidi RPC without holding a stream open across calls (same simplification the Python
     * client makes in `_grpc.py`'s `ingest()`: a long-lived streaming session across many pushes
     * is future work, not needed for `push()`'s surface). */
    suspend fun ingest(sourceName: String, rows: List<Row>, idempotencyKey: String?, partial: Boolean): IngestAckResult {
        val request = IngestRequest.newBuilder()
            .setSourceName(sourceName)
            .addAllRows(rows.map { it.toStruct() })
            .setPartial(partial)
            .setIdempotencyKey(idempotencyKey ?: "")
            .build()
        val ack = ingestStub.ingest(flowOf(request), authHeaders()).first()
        return IngestAckResult(
            outcome = ack.outcome.name,
            accepted = ack.accepted,
            dropped = ack.dropped,
            invalid = ack.invalid,
            retryAfterMs = ack.retryAfterMs,
            error = ack.error,
            rowErrors = ack.rowErrorsList,
            duplicate = ack.duplicate,
            replayed = ack.replayed,
        )
    }
}

File: clients/typescript/src/grpc-transport.ts
/**
 * Tier 1 gRPC transport (design doc §3): StreamService.SubscribeTable for deltas,
 * TableService.Rows/Search/Validate/List for the catalog and snapshot, and the bidi
 * IngestService.Ingest for pushes. Plaintext h2c by default, prior knowledge -- no TLS
 * negotiation, matching how the engine is actually run from source (§3.2's --urls trap). A
 * `https://` target switches to ALPN-negotiated h2 over TLS (the server side of this is
 * `--Tls:Enabled true` on the host, see `tools/tls/dev-cert.sh`) -- see `parseGrpcTarget` below.
 *
 * Node-only, and deliberately never imported eagerly: index.ts reaches this module only via a
 * dynamic `import()` gated on a Node-runtime check, so a browser bundle that imports
 * `@streamsforge/client` never pulls in `@grpc/grpc-js` (a browser cannot speak h2c gRPC at all --
 * see the package README).
 *
 * Proto loaded dynamically via @grpc/proto-loader (no generated stubs to keep in sync, unlike the
 * Python client's checked-in `_pb/streamsforge_pb2*.py`) -- `keepCase: false` gives camelCase
 * field names matching the REST DTOs' own convention (types.ts), and `longs: 'String'` avoids
 * silently truncating an int64 seq/weight to a JS double.
 *
 * Row payloads travel as google.protobuf.Struct; proto-loader represents one as its own raw wire
 * shape (`{fields: {name: {kind: 'stringValue', stringValue: ...}}}`), not a plain object --
 * structToRow()/rowToStruct() below are the "typing" story (design doc §2: rows stay dicts, a
 * DataFrame/consumer re-types them anyway). A number that doesn't round-trip through Struct's
 * IEEE-754 double (an int64 beyond 2**53) is a documented, not fixed, edge -- nothing in the
 * reference demo crosses it.
 */

import * as grpc from "@grpc/grpc-js";
import * as protoLoader from "@grpc/proto-loader";
import { fileURLToPath } from "node:url";
import path from "node:path";
import { StreamsForgeError } from "./errors.js";
import type { IngestAckDto } from "./ingest.js";
import type { Transport } from "./transport.js";
import type { Delta, Row } from "./zset.js";

const PROTO_PATH = path.join(path.dirname(fileURLToPath(import.meta.url)), "proto", "streamsforge.proto");

// ---- Struct <-> Row conversion -------------------------------------------------------------

interface ProtoValue {
  kind?: string;
  nullValue?: unknown;
  numberValue?: number;
  stringValue?: string;
  boolValue?: boolean;
  structValue?: ProtoStruct;
  listValue?: { values?: ProtoValue[] };
}
interface ProtoStruct {
  fields?: Record<string, ProtoValue>;
}

function valueToJs(v: ProtoValue): unknown {
  switch (v.kind) {
    case "nullValue":
      return null;
    case "numberValue":
      return v.numberValue;
    case "stringValue":
      return v.stringValue;
    case "boolValue":
      return v.boolValue;
    case "structValue":
      return structToRow(v.structValue);
    case "listValue":
      return (v.listValue?.values ?? []).map(valueToJs);
    default:
      return null;
  }
}

function structToRow(struct: ProtoStruct | undefined): Row {
  const row: Row = {};
  for (const [k, v] of Object.entries(struct?.fields ?? {})) row[k] = valueToJs(v);
  return row;
}

function jsToValue(x: unknown): ProtoValue {
  if (x === null || x === undefined) return { kind: "nullValue", nullValue: "NULL_VALUE" };
  if (typeof x === "number") return { kind: "numberValue", numberValue: x };
  if (typeof x === "string") return { kind: "stringValue", stringValue: x };
  if (typeof x === "boolean") return { kind: "boolValue", boolValue: x };
  if (Array.isArray(x)) return { kind: "listValue", listValue: { values: x.map(jsToValue) } };
  return { kind: "structValue", structValue: rowToStruct(x as Row) };
}

function rowToStruct(row: Row): ProtoStruct {
  const fields: Record<string, ProtoValue> = {};
  for (const [k, v] of Object.entries(row)) fields[k] = jsToValue(v);
  return { fields };
}

// ---- proto loading (cached: loadSync parses + compiles descriptors, not free) ---------------

interface StreamsForgeV1Package {
  TableService: grpc.ServiceClientConstructor;
  StreamService: grpc.ServiceClientConstructor;
  IngestService: grpc.ServiceClientConstructor;
}

let cachedPackage: StreamsForgeV1Package | null = null;

function loadV1(): StreamsForgeV1Package {
  if (cachedPackage) return cachedPackage;
  const packageDef = protoLoader.loadSync(PROTO_PATH, {
    keepCase: false,
    longs: String,
    enums: String,
    defaults: true,
    oneofs: true,
  });
  const pkg = grpc.loadPackageDefinition(packageDef) as unknown as {
    streamsforge: { v1: StreamsForgeV1Package };
  };
  cachedPackage = pkg.streamsforge.v1;
  return cachedPackage;
}

// ---- wire DTOs (post proto-loader, camelCase, longs-as-string) ------------------------------

interface TableRowWire {
  row: ProtoStruct;
  weight: string;
}
interface TableDeltaBatchWire {
  tableName: string;
  seq: string;
  deltas: TableRowWire[];
}
interface TableDefinitionWire {
  id: string;
  name: string;
  [key: string]: unknown;
}
interface ValidateTableResponseWire {
  ok: boolean;
  diagnostics: Array<{ message: string; line: number; column: number; severity: string }>;
  planSummary?: string;
  streamInputs: string[];
  tableInputs: string[];
  outputSchema: Array<{ name: string; type: string }>;
}

function promisify<Req, Res>(
  fn: (req: Req, md: grpc.Metadata, cb: (err: grpc.ServiceError | null, res: Res) => void) => void,
  req: Req,
  md: grpc.Metadata,
): Promise<Res> {
  return new Promise((resolve, reject) => {
    fn(req, md, (err, res) => (err ? reject(err) : resolve(res)));
  });
}

// ---- target parsing (scheme decides plaintext vs TLS) -----------------------------------

export interface ParsedGrpcTarget {
  /** Bare `host:port`, scheme stripped -- what `@grpc/grpc-js`'s client constructors expect. */
  target: string;
  /** Whether to dial with `createSsl` (true) or `createInsecure` (false). */
  tls: boolean;
}

/**
 * `grpc=` (ConnectOptions.grpc / STREAMSFORGE_GRPC) may be a bare `host:port` (plaintext,
 * unchanged since before TLS support), an explicit `http://host:port` (also plaintext), or
 * `https://host:port` (TLS). Anything else with no recognized scheme is treated as bare
 * `host:port` -- the historical default, so an existing caller's target string keeps working
 * unmodified.
 */
export function parseGrpcTarget(raw: string): ParsedGrpcTarget {
  if (raw.startsWith("https://")) return { target: raw.slice("https://".length), tls: true };
  if (raw.startsWith("http://")) return { target: raw.slice("http://".length), tls: false };
  return { target: raw, tls: false };
}

export interface GrpcTransportOptions {
  /** PEM text (not a path -- callers resolve a file path before reaching here, see index.ts's
   * `resolveCa`). Passed to `createSsl`'s `rootCerts`; omitted means "trust the system roots",
   * which is what a certificate from a real CA needs and a self-signed dev cert does not have. */
  ca?: string;
  /** false = accept any certificate (self-signed/invalid), the gRPC equivalent of http.ts's
   * `verify: false` -- dev-only, see this option's doc comment on ConnectOptions in index.ts. */
  verify?: boolean;
}

// ---- Transport ---------------------------------------------------------------------------

export class GrpcTransport implements Transport {
  readonly name = "grpc";
  private readonly tables: grpc.Client & Record<string, unknown>;
  private readonly stream: grpc.Client & Record<string, unknown>;
  private readonly ingestClient: grpc.Client & Record<string, unknown>;

  constructor(
    target: string,
    private readonly getToken: () => Promise<string>,
    opts: GrpcTransportOptions = {},
  ) {
    const v1 = loadV1();
    const { target: bareTarget, tls } = parseGrpcTarget(target);
    let creds: grpc.ChannelCredentials;
    if (tls) {
      const rootCerts = opts.ca ? Buffer.from(opts.ca, "utf-8") : null;
      // grpc-js's VerifyOptions (confirmed against the installed @grpc/grpc-js@1.14.4) supports
      // both rejectUnauthorized and checkServerIdentity -- setting both is belt-and-braces (some
      // older grpc-js releases only honored the latter for the "wrong hostname" half of
      // verification, not "untrusted issuer").
      const verifyOptions = opts.verify === false ? { rejectUnauthorized: false, checkServerIdentity: () => undefined } : undefined;
      creds = grpc.credentials.createSsl(rootCerts, null, null, verifyOptions);
    } else {
      creds = grpc.credentials.createInsecure();
    }
    this.tables = new v1.TableService(bareTarget, creds) as unknown as grpc.Client & Record<string, unknown>;
    this.stream = new v1.StreamService(bareTarget, creds) as unknown as grpc.Client & Record<string, unknown>;
    this.ingestClient = new v1.IngestService(bareTarget, creds) as unknown as grpc.Client & Record<string, unknown>;
  }

  private async metadata(): Promise<grpc.Metadata> {
    const md = new grpc.Metadata();
    md.add("authorization", `Bearer ${await this.getToken()}`);
    return md;
  }

  /** Proves the channel AND the JWT actually work -- called by connect()'s gRPC attempt and its
   * "auto" fallback path (index.ts), mirroring the Python client's own `list_tables(timeout=3.0)`
   * probe. */
  async listTables(): Promise<TableDefinitionWire[]> {
    const md = await this.metadata();
    const call = (this.tables as unknown as {
      List: (req: unknown, md: grpc.Metadata, cb: (err: grpc.ServiceError | null, res: { tables: TableDefinitionWire[] }) => void) => void;
    }).List;
    const resp = await promisify(call.bind(this.tables), {}, md);
    return resp.tables;
  }

  private async resolveTableId(name: string): Promise<string> {
    for (const t of await this.listTables()) {
      if (t.name === name) return t.id;
    }
    throw new StreamsForgeError(`no such table '${name}'`);
  }

  async validate(sql: string): Promise<ValidateTableResponseWire> {
    const md = await this.metadata();
    const call = (this.tables as unknown as {
      Validate: (req: unknown, md: grpc.Metadata, cb: (err: grpc.ServiceError | null, res: ValidateTableResponseWire) => void) => void;
    }).Validate;
    return promisify(call.bind(this.tables), { sql }, md);
  }

  async search(tableName: string, query: string, limit = 50): Promise<Row[]> {
    const id = await this.resolveTableId(tableName);
    const md = await this.metadata();
    const call = (this.tables as unknown as {
      Search: (req: unknown, md: grpc.Metadata, cb: (err: grpc.ServiceError | null, res: { rows: TableRowWire[] }) => void) => void;
    }).Search;
    const resp = await promisify(call.bind(this.tables), { id, query, limit }, md);
    return resp.rows.filter((r) => Number(r.weight) > 0).map((r) => structToRow(r.row));
  }

  // ---- Transport interface ----

  async snapshot(tableName: string, limit = 500): Promise<readonly [Delta[], number]> {
    const id = await this.resolveTableId(tableName);
    const md = await this.metadata();
    const call = (this.tables as unknown as {
      Rows: (req: unknown, md: grpc.Metadata, cb: (err: grpc.ServiceError | null, res: { rows: TableRowWire[]; seq: string }) => void) => void;
    }).Rows;
    const resp = await promisify(call.bind(this.tables), { id, limit, offset: 0 }, md);
    const deltas: Delta[] = resp.rows.map((r) => [structToRow(r.row), Number(r.weight)] as const);
    return [deltas, Number(resp.seq)] as const;
  }

  /**
   * Establishes the subscription -- awaits auth metadata, then CREATES the streaming call -- and
   * only THEN resolves, per Transport.subscribe()'s contract (see transport.ts's doc comment).
   * Creating the call is the real handshake here: @grpc/grpc-js puts the request on the wire as
   * soon as the method is invoked, independent of whether/when anything reads the response
   * stream (confirmed empirically against this engine -- a delta pushed after call-creation but
   * before the first read is still delivered once reading starts), so by the time this function
   * returns, the server has already registered the subscription. This function must NOT be an
   * `async function*` itself: a generator's body -- including this handshake -- would not run
   * until the caller's first `.next()`, silently deferring the handshake past whatever the
   * caller does next (in live-table.ts's case, the snapshot read) and reopening the exact race
   * this contract exists to close.
   */
  async subscribe(tableName: string, signal: AbortSignal): Promise<AsyncIterable<readonly [Delta[], number]>> {
    const md = await this.metadata();
    const subscribeFn = (this.stream as unknown as {
      SubscribeTable: (req: unknown, md: grpc.Metadata) => grpc.ClientReadableStream<TableDeltaBatchWire>;
    }).SubscribeTable;
    const call = subscribeFn.call(this.stream, { name: tableName }, md);

    // grpc-js's Call.cancel() is documented safe from any context (unlike a bare async
    // generator's own .close(), which is a footgun only in threaded runtimes -- moot in JS, but
    // AbortSignal is still the cleanest way to tie this stream's lifetime to live-table.ts's).
    const onAbort = () => call.cancel();
    if (signal.aborted) {
      // Establishment raced a close() that landed first -- cancel immediately rather than
      // leaking a half-open call nothing will ever read from.
      call.cancel();
    } else {
      signal.addEventListener("abort", onAbort);
    }

    const iterate = async function* (): AsyncGenerator<readonly [Delta[], number]> {
      try {
        for await (const batch of call as unknown as AsyncIterable<TableDeltaBatchWire>) {
          const deltas: Delta[] = batch.deltas.map((d) => [structToRow(d.row), Number(d.weight)] as const);
          yield [deltas, Number(batch.seq)] as const;
        }
      } catch (err) {
        if (signal.aborted) return; // an intentional cancel() surfaces as a CANCELLED error -- not a real failure
        throw new StreamsForgeError(`gRPC SubscribeTable('${tableName}') stream ended: ${String(err)}`);
      } finally {
        signal.removeEventListener("abort", onAbort);
      }
    };
    return { [Symbol.asyncIterator]: () => iterate() };
  }

  // ---- ingest ----

  /** One request, one ack, over a fresh bidi stream -- real backpressure semantics from the bidi
   * RPC (the server does not ack until PushAsync returns) without holding a stream open across
   * calls. A long-lived streaming session (sustained backpressure across many pushes) is future
   * work, not needed for this client's push() surface -- mirrors _grpc.py's own ingest(). */
  async ingest(sourceName: string, rows: Row[], idempotencyKey: string | undefined, partial: boolean): Promise<IngestAckDto> {
    const md = await this.metadata();
    const ingestFn = (this.ingestClient as unknown as {
      Ingest: (md: grpc.Metadata) => grpc.ClientDuplexStream<unknown, IngestAckDto>;
    }).Ingest;
    const call = ingestFn.call(this.ingestClient, md);
    return new Promise((resolve, reject) => {
      let settled = false;
      call.on("data", (ack: IngestAckDto) => {
        settled = true;
        resolve(ack);
        call.end();
      });
      call.on("error", (err: Error) => {
        if (!settled) reject(new StreamsForgeError(`gRPC Ingest('${sourceName}') failed: ${err.message}`));
      });
      call.on("end", () => {
        if (!settled) reject(new StreamsForgeError(`gRPC Ingest('${sourceName}') stream closed with no ack`));
      });
      call.write({
        sourceName,
        rows: rows.map(rowToStruct),
        partial,
        idempotencyKey: idempotencyKey ?? "",
      });
    });
  }

  close(): void {
    this.tables.close();
    this.stream.close();
    this.ingestClient.close();
  }
}
